### DISCUSSION TITLE
get_zigbee_cluster_attribute

### DISCUSSION THREAD

**Discussion by juhanjuku** [zigpy/zigpy#859] [Ideas]
We have service to **set** any cluster/attribute in HA. We can also **read** any cluster/attribute from 'Manage Clusters' page.
**How difficult would be to add also zha.get_zigbee_cluster_attribute service?**
With such service could read and use in automations also attributes which are not polled to zigbee.db or not refreshed properly.
There are several discussions about this need in HA community.

**Comment by dmulcahey** [comment]
What are real world use cases? Chances are we aren’t configuring something correctly and that should change as opposed to doing this.

**Reply by juhanjuku** [reply]
I have now problem with several smart plugs where smartenergy_metering_summation entity updates only once after restart.
https://github.com/home-assistant/core/issues/59081
With get_zigbee_cluster_attribute service I could force this entity update.

![pic0](https://user-images.githubusercontent.com/12792066/144495652-a59ed5b8-cd87-478d-ba3c-cb7406d66ebf.png)

Other users are also asking for such service
[(https://community.home-assistant.io/t/zha-read-cluster-attributes-programatically/277958)](https://community.home-assistant.io/t/zha-read-cluster-attributes-programatically/277958)

I'm experimenting with quirks had idea to scan clusters/attribues on new device with such service.
I think this service would be logical addition to set_zigbee_cluster_attribute

 Actually I was just asking developers opinion how big work it would be to add this service.
Unfortunately my own programming skills are not on needed level :(

**Reply by dmulcahey** [reply]
> I have now problem with several smart plugs where smartenergy_metering_summation entity updates only once after restart. [home-assistant/core#59081](https://github.com/home-assistant/core/issues/59081) With get_zigbee_cluster_attribute service I could force this entity update.
> 

This is what I meant… something is broken here if that’s the case… we don’t want folks leveraging services to hack around things we need to fix

**Reply by dmulcahey** [reply]
We added these sensors very recently… have you tried reconfiguring each device? I believe we had a note about this being necessary…

**Reply by juhanjuku** [reply]
Yes, I have tried remove/add but no help.
I'm surprised, why it's bad when users have some temporary ways, until devs come up with proper solution.
We all see how overloaded you devs are and we wait  in que with our problems, but why we can't help ourselvs with temporary hacks?

**Reply by puddly** [reply]
It's exposed over the websocket interface so there's always that (untested):

```shell
$ hass-cli raw ws zha/devices/clusters/attributes/value --json='{"ieee": "00:12:4b:00:1c:a1:b8:46", "endpoint_id": 1, "cluster_id": 123, "cluster_type": "in", "attribute": 456}'
```

**Reply by Adminiuga** [reply]
Iirc some plug may use none standard attribute type and fail attribute reporting configuration for summation. May need quirk to redefine attribute with the type the plug expects.

**Reply by Adminiuga** [reply]
The homeassistant.update_entity service should read the attributes of the channel, no?

**Reply by juhanjuku** [reply]
Tested right now homeassistant.update_entity service from Developer Tools - unfortunately it does not refresh freezed entities

**Reply by juhanjuku** [reply]
> > I have now problem with several smart plugs where smartenergy_metering_summation entity updates only once after restart. [home-assistant/core#59081](https://github.com/home-assistant/core/issues/59081) With get_zigbee_cluster_attribute service I could force this entity update.
> 
> This is what I meant… something is broken here if that’s the case… we don’t want folks leveraging services to hack around things we need to fix
@dmulcahey  Why it's bad when users have some temporary ways, until devs come up with proper solution.
We all see how overloaded you devs are and we wait in que with our problems, but why we can't help ourselvs with temporary hacks?

**Reply by lodegaard** [reply]
I would very much like this feature aswell.

I have some Zigbee devices in my network that has vendor specific attributes that are not reported periodically (it could be general attributes aswell). Since there is no easy way (that I know of) to set up custom attribute reporting for devices, polling is probably the way to go.

**Reply by Ashutor** [reply]
I have been using the solution from @antoineDion for more than a year. Its listed here - https://github.com/home-assistant/core/issues/44032.

It breaks once in a while when the ZHA component is updated, but then I just readd the tweaks as a ZHA_custom-component and voila - back in business. I then read the attribute using an automation.

**Reply by sersol-ai** [reply]
I would like to have get_zigbee_cluster_attribute function - I need read cluster attributes manually on IKEA FYRTUR every time  every time after replacing batteries to fix position and battery reading.  It seem code is already there https://github.com/home-assistant/core/issues/44032

**Reply by BAAAASS** [reply]
I too am looking for this feature.

I have an old-school outdoor sun-shade that open with a modern switch.  The basic functionality works perfectly, except that when I close it, it never goes back to physically perfectly closed.  It seems to almost always stay open 5% (physically), but according to the switch it is on 0%.  So the switch thinks its closed, but when I look at the awning, it is still open by a small bit.

I really don't want to spend $5000 to replace the shade and motor etc. to fix this "properly", so I really need help to get it work practically.

For now, I have been going into the cluster settings manually and updating the value, but it would be much easier to automate a triger once it has been opened, just to increase the value of open% by 5% or 6%.
I.e. if it is currently opned to 60%, change the value to 63%.  Then when I close the shade, it should automatically close completely.

However, in order for me to increase the value, I need to be able to read it in programatically (as part of the script).

**Reply by Ashutor** [reply]
@BAAAASS I have found that HACS addon zha_toolkit allows you to read cluster attributes directly using the zha_toolkit.attr_read function. I call it using a HA script and then parse the variable returned. There are some well documented examples that should work for your use case - https://github.com/mdeweerd/zha-toolkit.

**Comment by jeffmckee** [comment]
I have an issue with this too. My Phillips SML01 Motion Sensors don't update illuminance or temperature automatically, unless I read the attribute, so I'm hoping to create a time pattern automation to do this.

**Reply by puddly** [reply]
Reconfigure the device from ZHA, attribute reporting was likely not set up when the device joined.

**Comment by Zaleo80** [comment]
Now that I am picking up some quirky Zigbee devices I am missing this action as well (Services are now called actions) before I used ZHA Toolkit, but the zha_toolkit.attr_read "service" also seems to be missing.

**Reply by Ashutor** [reply]
There is an issue (and a fix in beta 1.1.19) listed for this under ZHA_Toolkit - https://github.com/mdeweerd/zha-toolkit/issues/248.

**Reply by Zaleo80** [reply]
> There is an issue (and a fix in beta 1.1.19) listed for this under ZHA_Toolkit - [mdeweerd/zha-toolkit#248](https://github.com/mdeweerd/zha-toolkit/issues/248).

Solved indeed, still a +1 for getting this functionality "action/service" in HA Core.
